### ISSUE TITLE
Suppressing warnings

### ISSUE THREAD

**Issue by Hi-Angel** [purescript/purescript#4557]
## Summary

Sometimes there are undesirable warnings that the compiler prints anyway. For example, most PureScript projects break their internals to multiple files for their own reasons, but for the end-user this is nothing but a nuisance, so you typically want to import all project files by just the project name. But then the compiler starts spewing `ImplicitQualifiedImport` on all of your warnings.

In practice what happens then is that useful warnings get lost among those unwanted ones, so you don't know when there's actually something to fix in the code.

Related old discussion [here](https://discourse.purescript.org/t/suppressing-warnings-in-code/2485/12)

## Motivation

Being able to turn warnings into errors, which is impossible to do when there are warnings you'd prefer not to fix.

## Proposal

The compiler should provide an option similar to GCC's `-Wno-FOO` to suppress warning `FOO`.

## Examples

```
purs --Wno-ImplicitQualifiedImport Foo.purs
```

**Comment by wclr** [comment]
Have you seen https://github.com/natefaubion/purescript-psa?

**Comment by Hi-Angel** [comment]
> Have you seen https://github.com/natefaubion/purescript-psa?

Yes… Just as I was writing an answer, I thought: hmm, is it maybe possible to use with Spago…? Thing is, while looking through configs I don't see any way to override `purs` to `psa` for example.

**Comment by Hi-Angel** [comment]
*(but yeah, I find it odd that this functional is located elsewhere and not in the compiler. Remind me of the fact that [optimizations are also outsourced to a separate project](https://github.com/aristanetworks/purescript-backend-optimizer). I'm wondering if there's a reason this is happening.)*

**Comment by wclr** [comment]
> Thing is, while looking through configs I don't see any way to override purs to psa for example.

You can try to rename `purs -> smth`, and `psa -> purs`, and give `psa --purs smth` arg 😄

**Comment by garyb** [comment]
#3493 would probably be a prerequisite of this if we were to do it.

**Comment by garyb** [comment]
Well, some of the warning are not safe to suppress without knowledge of packages. All of the annoying import warnings aside from duplicate/unused are there to ensure that additions to the modules being imported will not cause scope conflicts.

If the compiler was module-aware it could be lax about those kind of warnings for all dependencies within a project, and only warn when referencing modules outside of the project. 

We made the decision to introduce those warnings rather than require every additive change to an existing module in a library to be released as breaking version.

I imagine we'd default to the equivalent of `psa`'s `--censor-lib` when the compiler is module-aware too, which isn't directly relevant to the issue at hand but is sorta related!

**Comment by wclr** [comment]
> Well, some of the warning are not safe to suppress without knowledge of packages. All of the annoying import warnings aside from duplicate/unused are there to ensure that additions to the modules being imported will not cause scope conflicts.

I think safe and unsafe in this context is kind of subjective, as well as desire to suppress warnings. I myself have a fault to be used to censor "ImplicitQualifiedImport".

To allow something like this in the code:
```
import Some.Foo as S
import Some.Bar as S
```
But to me this looks like a kind of code smell. Is it subjective? If things are placed in different namespaces `Some.Foo` and `Some.Bar` there must be a reason for this? 

> most PureScript projects break their internals to multiple files for their own reasons, 

There definitely should be reasons. And those reasons do (should) not differ much from the reasoning of the external library's authors, etc. If those things can be semantically treated as coming from just the same origin why are they not placed in the same module (namespace)? There must be some *reason*.

> but for the end-user this is nothing but a nuisance

I think this is a bit negligent attitude to the issue.

To shorten, one can use something like this:
```
import Some.Foo as SF
import Some.Bar as SB
```

That's a old discuss, though I repeat it again. I hope something like this https://github.com/purescript/purescript/issues/3493 never gets landed as a way to introduce/require this change to imports:

```
import "effect" Effect
import "some-cute-library" Default
```

This would be a bad thing to happen to the language. 

Though I believe the compiler could have some benefits of introducing awareness of some light-weight packaging system (for example to allow internal modules not exposed modules for example), but not introduce an additional required namespace level for the modules.

People should appreciate the value of module names as namespaces, and treat and use them correctly (with respect to the ecosystem) not cause conflicts and ambiguities, the same attitude should be for their project's "internals".

**Comment by Hi-Angel** [comment]
> There definitely should be reasons. And those reasons do (should) not differ much from the reasoning of the external library's authors, etc. If those things can be semantically treated as coming from just the same origin why are they not placed in the same module (namespace)? There must be some _reason_.
> 
> > but for the end-user this is nothing but a nuisance
> 
> I think this is a bit negligent attitude to the issue.

I think most frequent reason authors do that is same one any developer breaks their code to different modules: it's more convenient to maintain when you have components logically separated. But a library user, as opposed to the original library developer, doesn't use **everything** from the library, they just use 2-3 functions from `Lib.Module1` and 1 function from `Lib.Module2`. To me it just doesn't make sense to force people to remember that certain function was located not just in `Lib`, but exactly in the ModuleX.

More rarely of course the reason is names clash, but IIRC there's a "name is shadowed" warning for this, so the compiler should catch that.

**Comment by Hi-Angel** [comment]
> Well, some of the warning are not safe to suppress without knowledge of packages. All of the annoying import warnings aside from duplicate/unused are there to ensure that additions to the modules being imported will not cause scope conflicts.

@garyb unless I'm missing something, this already works. See, given this code:

* `src/Module1.purs`
    ```haskell
    module Module1 where

    import Prelude

    foo = identity
    ```
* `src/Module2.purs`
    ```haskell
    module Module1 where

    import Prelude

    foo = identity
    ```
* `src/Main.purs`
    ```haskell
    module Main where

    import Prelude

    import Effect (Effect)
    import Module1 as M
    import Module2 as M

    bar = M.foo

    main :: Effect Unit
    main = pure unit
    ```

Running `spago build` gives me:

```haskell
[1 of 1] Compiling Main
Error found:
in module Main
at src/Main.purs:9:7 - 9:12 (line 9, column 7 - line 9, column 12)

  Conflicting definitions are in scope for value foo from the following modules:

    Module1
    Module2
```

**Comment by garyb** [comment]
Indeed, that's the error that the warnings protect against occurring in the future when your dependencies have additive (minor version number bump) changes.

**Comment by wclr** [comment]
> Indeed, that's the error that the warnings protect against occurring in the future when your dependencies have additive (minor version number bump) changes.

I thought by "safe" yout meant ephemeral feeling of control of user's modules apis)

I may be too rigid to say that qualifying multiple models with the same name is obviously a bad thing. Actually if it fits the context it is applicable, and it is ok to suppress this kind of warnings also. It would not be difficult to add a warning censor option to the compiler, if it is really needed.

**Comment by bakhtiyarneyman** [comment]
I want downstream users to be able to suppress warnings inline on case-by-case basis using pragmas (similar to Rust's `#[allow(...)]` macros). Not having this feature makes me hesitant to use `class Warn` from `Prim.TypeError`.

**Comment by f-f** [comment]
@bakhtiyarneyman this can be done with spago@next as well, look for `censorLibraryWarnings` in the docs
